mypyc/test-data/run-misc.test

```diff
@@ -84,6 +84,33 @@ assert f(a) is a
 assert g(None) == 1
 assert g(a) == 2
 
+[case testInferredOptionalAssignment]
+from typing import Any, Generator
+
+def f(b: bool) -> Any:
+    if b:
+        x = None
+    else:
+        x = 1
+
+    if b:
+        y = 1
+    else:
+        y = None
+
+    m = 1 if b else None
+    n = None if b else 1
+
+def gen(b: bool) -> Generator[Any, None, None]:
+    if b:
+        y = 1
+    else:
+        y = None
+    yield y
+
+f(False)
+next(gen(False))
```

Comment:
Try calling with true arguments as well, to make sure both code paths work when compiled?

Test the return value of `next(...)`?